File: languages/scala/src/main/scala/fp/scala/exampleapp/forms/OrderForm.scala
package fp.scala.exampleapp.forms

import scala.util.parsing.json.JSON

case class OrderForm(email: String = "",
                      addressLine1: String = "",
                      city: String = "",
                      state: String = "",
                      zip: String = "",
                      date: String = "",
                      sku: String = "",
                      quantity: String = "",
                      validationErrors: List[String] = List())  {


  def values = List(email, addressLine1, city, state, zip, date, sku, quantity)

  def validate(): OrderForm = {
    this match {
      case f if !values.filter(isBlank).isEmpty =>
        f.copy(validationErrors = List("All fields are required"))
      case f => f
    }
  }

  def isValid() = validate().validationErrors.isEmpty

  private def isBlank(s:String):Boolean = {s == null || s.trim == ""}
}

object OrderForm {
  def fromJson(json: String): OrderForm = {
    JSON.parseFull(json) match {
      case Some(parsed:Map[String, Any]) => fromMap(parsed).validate()
      case _ => OrderForm()
    }
  }

  def fromMap(map: Map[String, Any]): OrderForm = {
    val stringMap: Map[String, String] = map.mapValues(_.toString)
    
    OrderForm(stringMap("email"),
      stringMap("address"),
      stringMap("city"),
      stringMap("state"),
      stringMap("zip"),
      stringMap("date"),
      stringMap("sku"),
      stringMap("quantity"))
  }

}

File: languages/java/src/main/java/fp/java/exampleapp/forms/OrderForm.java
package fp.java.exampleapp.forms;

import org.codehaus.jackson.map.ObjectMapper;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class OrderForm {
    private String email = "";
    private String address = "";
    private String city = "";
    private String state = "";
    private String zip = "";
    private String date = "";
    private String sku = "";
    private String quantity = "";
    private List<String> validationErrors = new ArrayList<>();
    private static ObjectMapper mapper = new ObjectMapper();
    
    public static OrderForm fromJson(String json) throws IOException {
        return mapper.readValue(json, OrderForm.class);
    }

    public void validate()  {
        validationErrors.clear();
        if(isBlank(email) || isBlank(address) || isBlank(city) ||
                isBlank(state) || isBlank(zip) || isBlank(date) || isBlank(sku) ||
                isBlank(quantity)){
            validationErrors.add("All fields are required");
        }
    }

    public Boolean isValid() {
        validate();
        return validationErrors.isEmpty();
    }

    private Boolean isBlank(String s) {return s == null || s.trim().equals("");}

    public String getEmail() {
        return email;
    }

    public void setEmail(String email) {
        this.email = email;
    }

    public String getAddress() {
        return address;
    }

    public void setAddress(String address) {
        this.address = address;
    }

    public String getCity() {
        return city;
    }

    public void setCity(String city) {
        this.city = city;
    }

    public String getState() {
        return state;
    }

    public void setState(String state) {
        this.state = state;
    }

    public String getZip() {
        return zip;
    }

    public void setZip(String zip) {
        this.zip = zip;
    }

    public String getDate() {
        return date;
    }

    public void setDate(String date) {
        this.date = date;
    }

    public String getSku() {
        return sku;
    }

    public void setSku(String sku) {
        this.sku = sku;
    }

    public String getQuantity() {
        return quantity;
    }

    public void setQuantity(String quantity) {
        this.quantity = quantity;
    }

    public List<String> getValidationErrors() {
        return validationErrors;
    }

    public void setValidationErrors(List<String> validationErrors) {
        this.validationErrors = validationErrors;
    }
}
